gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java

```diff
@@ -1334,47 +1354,32 @@ public void putAssignmentsAndCategoryItemsInMatrix(Map<String, GbStudentGradeInf
 
 						boolean gradeable = false;
 
-						for (final PermissionDefinition permission : permissions) {
-							// we know they are all GRADE so no need to check here
-
-							boolean categoryOk = false;
-							boolean groupOk = false;
+						//If TA does not have all the permissions over categories and groups
+						if (!allPermissions) {
 
-							final Long permissionCategoryId = permission.getCategoryId();
-							final String permissionGroupReference = permission.getGroupReference();
+							//Check category and group permissions
+							List<String> arr = permByCat.get(gradeCategoryId) != null ? permByCat.get(gradeCategoryId) : new ArrayList<String>();
 
-							log.debug("permissionCategoryId: {}", permissionCategoryId);
-							log.debug("permissionGroupReference: {}", permissionGroupReference);
-
-							// if permissions category is null (can grade all categories) or they match (can grade this category)
-							if (!categoriesAreEnabled() || (permissionCategoryId == null || permissionCategoryId.equals(gradeCategoryId))) {
-								categoryOk = true;
-								log.debug("Category check passed");
+							//add groups with permission in all categories
+							if (permByCat.containsKey(Long.valueOf(-1L))) {
+								arr.addAll(permByCat.get(Long.valueOf(-1L)));
 							}
 
-							// if group reference is null (can grade all groups) or group membership contains student (can grade this group)
-							if (StringUtils.isBlank(permissionGroupReference)) {
-								groupOk = true;
-								log.debug("Group check passed #1");
-							} else {
-								final List<String> groupMembers = groupMembershipsMap.get(permissionGroupReference);
-								log.debug("groupMembers: {}", groupMembers);
+							for (String group:(List<String>) arr) {
```

Comment:
You need to space this out as it is hard to read. Also, you don't need  to cast arr to a List<String> - it's already that.